### PR TITLE
Bug 1907628: OpenStack Documentation for MachineSets with Multiple Networks

### PR DESCRIPTION
**Pull request by iamemilio**
Due to the addition of the PrimarySubnets variable in the providerSpec
we wanted to ensure that users knew it existed and of its pitfalls.

Fixes: [OSASINFRA-2088](https://issues.redhat.com/browse/OSASINFRA-2088)

/label platform/openstack
/cc @maxwelldb

### CODE REVIEW

File: docs/user/openstack/README.md [hunk 1]
```diff
@@ -444,7 +444,7 @@ rm -rf ostest/
 
 Groups of Compute nodes are managed using the [MachineSet][machine-set-code] resource. It is possible to create additional MachineSets post-install, for example to assign workloads to specific machines.
 
-When running on OpenStack, the MachineSet has platform-specific fields under `spec.template.spec.providerSpec.value`:
+When running on OpenStack, the MachineSet has platform-specific fields under `spec.template.spec.providerSpec.value`. More details about what values can be set in the `providerSpec` can be found in the [API definition](provider-spec-definition).
```
Comment (maxwelldb) [inline docs/user/openstack/README.md:447]:
```suggestion
When running on OpenStack, the MachineSet has platform-specific fields under `spec.template.spec.providerSpec.value`. For more information about the values that you can set in `providerSpec`, see [the API definition](provider-spec-definition).
```

File: docs/user/openstack/README.md [hunk 2]
```diff
@@ -501,6 +501,12 @@ spec:
           availabilityZone: <optional_openstack_availability_zone>
 ```
 
+[provider-spec-definition]: https://github.com/openshift/cluster-api-provider-openstack/blob/155384b859c5b2fb5b7f11c9111d3f8e4f3066bd/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L31
+
+#### Defining a New MachineSet With Multiple Networks
```
Comment (maxwelldb) [inline docs/user/openstack/README.md:506]:
```suggestion
#### Defining a New MachineSet That Uses Multiple Networks
```

File: docs/user/openstack/README.md [hunk 3]
```diff
@@ -501,6 +501,12 @@ spec:
...
+
+When attaching multiple networks to a machineset, the primarySubnet in the `providerSpec` need to be set to the `machines subnet` or the node will be unable to join the cluster. The primary subnet is where a nodes kuberentes endpoints are published. Once this is properly set, you can should add all of the networks that you want attached to the machines in the `Networks` list in the `providerSpec`. Note that the network that the primary subnet is part of should also be added to this list.
```
Comment (maxwelldb) [inline docs/user/openstack/README.md:508]:
```suggestion
To attach multiple networks to a machine set, the `primarySubnet` value in `providerSpec` must be set to the `machines subnet`. If the value is not set, the node cannot join the cluster. 

Nodes' Kubernetes endpoints are published on the primary subnet. After you set the subnet, add all of the networks that you want to attach to your machines to the `Networks` list in `providerSpec`. You must also add the network that the primary subnet is part of to this list. 
```

Not 100% about what you mean by `machines subnet` here.

**Review by maxwelldb — APPROVED** [review]
TY for the notice. Comments added.

File: docs/user/openstack/README.md (hunk 3 above)
Comment (iamemilio) [reply docs/user/openstack/README.md:508]: I will update with a link to the definition. Chances are someone else will also be confused by that

Comment (maxwelldb) [reply docs/user/openstack/README.md]: Ah! `machines subnet` vs. `machinesSubnet`.  👍

File: docs/user/openstack/README.md [hunk 4]
```diff
@@ -501,6 +501,14 @@ spec:
           availabilityZone: <optional_openstack_availability_zone>
 ```
 
+[provider-spec-definition]: https://github.com/openshift/cluster-api-provider-openstack/blob/155384b859c5b2fb5b7f11c9111d3f8e4f3066bd/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L31
+
+#### Defining a New MachineSet That Uses Multiple Networks
+
+To attach multiple networks to a machineSet, the `primarySubnet` value in the `providerSpec` must be set to the [machines subnet](./customization.md#cluster-scoped-properties). If this value is not set correctly, the nodes will be unable to join the cluster.
+
+Nodes' Kubernetes endpoints are published on the primary subnet. After you set the subnet, add all of the networks that you want to attach to your machines to the `Networks` list in `providerSpec`. You must also add the network that the primary subnet is part of to this list.
```
Comment (EmilienM) [inline docs/user/openstack/README.md:510]: Is ```Nodes'``` added on purpose?

Comment (maxwelldb) [reply docs/user/openstack/README.md:510]: "The Kubernetes endpoints of nodes are published..." ?

Comment (iamemilio) [reply docs/user/openstack/README.md:510]: Yeah, The node's kubernetes endpoints are being published on the primary subnet

File: docs/user/openstack/customization.md [hunk 5]
```diff
@@ -150,9 +150,9 @@ platform:
...
+Before running the installer with a custom `machinesSubnet`, there are a few things you should check. First, make sure that you have a network and subnet available. Note that the installer will create ports on this network, and some ports will be given a fixed IP addresses in the subnet's CIDR. Make sure that the credentials that you give the installer are authorized to do this! Also note that the installer will not create a router, so make sure to do this if you need a router to attach floating IP addresses to the network you want to install the nodes onto. For now, this feature does not fully support provider networks.
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:153]: This and the following line would probably be more digestible as lists.

File: docs/user/openstack/README.md [hunk 6]
```diff
@@ -501,6 +501,14 @@ spec:
           availabilityZone: <optional_openstack_availability_zone>
 ```
 
+[provider-spec-definition]: https://github.com/openshift/cluster-api-provider-openstack/blob/155384b859c5b2fb5b7f11c9111d3f8e4f3066bd/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L31
+
+#### Defining a New MachineSet That Uses Multiple Networks
```
Comment (adduarte) [inline docs/user/openstack/README.md:506]: do you need the word "New"? Defining usually means making the first of something.

File: docs/user/openstack/README.md [hunk 7]
```diff
@@ -501,6 +501,14 @@ spec:
...
+
+To attach multiple networks to a machineSet, the `primarySubnet` value in the `providerSpec` must be set to the [machines subnet](./customization.md#cluster-scoped-properties). If this value is not set correctly, the nodes will be unable to join the cluster.
```
Comment (adduarte) [inline docs/user/openstack/README.md:508]: here you use "machineSet" (lower case), but in L447 you use "MachineSet" is that correct?

File: docs/user/openstack/customization.md [hunk 8]
```diff
@@ -150,9 +150,14 @@ platform:
...
+Once that is ironed out, you can pass the ID of the subnet to the install config with the `machinesSubnet` option. This subnet will be set as the primarySubnet in the `machineSpec` for all nodes in this cluster. It is also where the Kubernetes endpoints of the nodes in your cluster will be published, and where the VIP ports will be created. This does add some complexity to the install process that needs to be accounted for.
```
Comment (adduarte) [inline docs/user/openstack/customization.md:155]: "the install config" -> in the install-config.yaml file

Comment (adduarte) [inline docs/user/openstack/customization.md:155]: Do you need "This does add some complexity to the install process that needs to be accounted for."
if so, you might want to explain what the "complexity" is 
Or is the complexity the lines below?

File: docs/user/openstack/customization.md [hunk 9]
```diff
@@ -150,9 +150,14 @@ platform:
...
+
+1. The `machinesSubnet` must have DHCP enabled.
+2. The CIDR of `machinesSubnet` must match the cidr of `networks.machineNetwork`.
+3. The API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To avoid other services taking the ports assigned to the api and ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the install config to addresses that are outside of the DHCP allocation pool.
```
Comment (adduarte) [inline docs/user/openstack/customization.md:159]: again "the install config" could be confusing to a new user. can you use the  "install-config.yaml file"

File: docs/user/openstack/README.md (hunk 7 above)
Comment (adduarte) [reply docs/user/openstack/README.md:508]: You mention `primarySubnet` in `providerSpec` in ./customization.md#cluster-scoped-properties. 
and it sounds as if you are saying that the user needs to set `primarySubnet` somewhere, but if i am reading this correctly you mean that the user needs to set the `machineSubnet` which in turn will set the `primarySubent` in the `providerspec` of the generated machineSet... right?  anyways. I think its a bit hard to follow.

**Review by adduarte — COMMENTED** [review]
looks good, but it does need some polishing and perhaps as max suggested, use lists to outline the steps that need to be taken.

File: docs/user/openstack/README.md (hunk 7 above)
Comment (iamemilio) [reply docs/user/openstack/README.md:508]: I tried cleaning up the sentence. Unfortunately, I cant clean up the UX. I agree that its unclear and I dont like it. Lets make an epic for that if we want to change it.

**Comment by iamemilio** [comment]
I re-worked those sections. @maxwelldb @adduarte is this clearer?

File: docs/user/openstack/customization.md [hunk 10]
```diff
@@ -150,9 +150,17 @@ platform:
...
+In the `install-config.yaml`, the `machinesSubnet` is the subnet where the Kubernetes endpoints of the nodes in your cluster will be published, and where the Ingress and API ports will be created. The installer creates a network and subnet for this purpose by default, however, you have the option to prevent the installer from creating those resources and overriding them with your own by setting the value of the `machinesSubnet` to the ID of an existing OpenStack subnet. To use this feature there are a few requirements:
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:153]: This is an improvement! I'd probably do something like this suggestion, but do what you like. 

```suggestion
In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published.  The Ingress and API ports are created on this subnet, too.

By default, the installer creates a subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the ID of an existing OpenStack subnet. 

To use your own subnet, your cluster must meet the following requirements:
```

Thoughts:

- Do we need to indicate "network and subnet," here, or is just "subnet" sufficient?
- Use "ID" or "UUID?"

File: docs/user/openstack/customization.md [hunk 11]
```diff
@@ -150,9 +150,17 @@ platform:
...
 
-Once that is ironed out, you can pass the ID of the subnet to the install config with the `machinesSubnet` option. This subnet will be tagged as the primary subnet for the cluster. The nodes, and VIP port will be created on this subnet. This does add some complexity to the install process that needs to be accounted for. First of all, the subnet must have DHCP enabled. Also, the CIDR of the custom subnet must match the cidr of `networks.machineNetwork`. Be aware that the API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To avoid other services taking the ports assigned to the api and ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the install config to addresses that are outside of the DHCP allocation pool. Lastly, note that setting `externalDNS` while setting `machinesSubnet` is invalid usage, since the installer will not modify your subnet. If you want to add a DNS to your cluster while using a custom subnet, add it to the subnet in openstack [like this](https://docs.openstack.org/neutron/rocky/admin/config-dns-res.html).
+* The `machinesSubnet` must have DHCP enabled
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:155]: Could punctuate these as sentences. 

Structure could also be grammatically simpler. For example:

"The subnet that is used by `machinesSubnet` has DHCP enabled."

File: docs/user/openstack/customization.md [hunk 12]
```diff
@@ -150,9 +150,17 @@ platform:
...
+* The CIDR of `machinesSubnet` must match the cidr of `networks.machineNetwork`
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:156]:
```suggestion
* The CIDR of `machinesSubnet` matches the CIDR of `networks.machineNetwork`
```

File: docs/user/openstack/customization.md [hunk 13]
```diff
@@ -150,9 +150,17 @@ platform:
...
+* The installer user must have permission to create ports on this network, including ports with fixed IP addresses
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:157]:
```suggestion
* The installer user has permission to create ports on this network, including ports that have fixed IP addresses
```

File: docs/user/openstack/customization.md [hunk 14]
```diff
@@ -150,9 +150,17 @@ platform:
...
+
+You should also be aware of the following details:
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:159]: I'd like a more specific word than "details" if possible. Limitations? Requirements?

File: docs/user/openstack/customization.md [hunk 15]
```diff
@@ -150,9 +150,17 @@ platform:
...
+
+* If you plan to install with floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`
+* The API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To avoid other services taking the ports assigned to the api and ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:162]:
```suggestion
* By default, the API and Ingress VIPs try to take the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```

File: docs/user/openstack/customization.md [hunk 16]
```diff
@@ -150,9 +150,17 @@ platform:
...
+* Setting `externalDNS` while setting `machinesSubnet` is invalid usage, since the installer will not modify your subnet. If you want to add a DNS to your cluster while using a custom subnet, add it to the subnet in openstack [like this](https://docs.openstack.org/neutron/rocky/admin/config-dns-res.html).
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:163]:
```suggestion
* You cannot use the `externalDNS` property at the same time as a custom `machinesSubnet`. If you want to add a DNS to your cluster while using a custom subnet, [add it to the subnet in OpenStack](https://docs.openstack.org/neutron/rocky/admin/config-dns-res.html).
```

**Review by maxwelldb — APPROVED** [review]
Added some comments and suggestions.

File: docs/user/openstack/customization.md (hunk 10 above)
Comment (iamemilio) [reply docs/user/openstack/customization.md:153]: It creates both a network and subnet, so its good to be clear about that. UUID might be better

File: docs/user/openstack/customization.md (hunk 14 above)
Comment (iamemilio) [reply docs/user/openstack/customization.md:159]: how does expectations sound?

Comment (iamemilio) [reply docs/user/openstack/customization.md:159]: Limitations works

File: docs/user/openstack/README.md [hunk 17]
```diff
@@ -501,6 +501,14 @@ spec:
           availabilityZone: <optional_openstack_availability_zone>
 ```
 
+[provider-spec-definition]: https://github.com/openshift/cluster-api-provider-openstack/blob/155384b859c5b2fb5b7f11c9111d3f8e4f3066bd/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L31
+
+#### Defining a MachineSet That Uses Multiple Networks
+
+To attach multiple networks to a MachineSet, the `primarySubnet` value in the `providerSpec` must be the same OpenStack subnet as the [machines subnet](./customization.md#cluster-scoped-properties) in the `install-config.yaml`. If this value is not set correctly, the nodes will be unable to join the cluster.
```
Comment (maxwelldb) [inline docs/user/openstack/README.md:508]: Not set "correctly," or not set so that it contains `primarySubnet`? 

Just "correctly" could imply that there are a variety of ways to screw up the value, with what's in the previous sentence being just one of them. 

(Could expand if that's actually the case.)

File: docs/user/openstack/customization.md [hunk 18]
```diff
@@ -150,9 +150,18 @@ platform:
...
 ## Custom Subnets
 
-Before running the installer with a custom `machinesSubnet`, there are a few things you should check. First, make sure that you have a network and subnet available. Note that the installer will create ports on this network, and some ports will be given a fixed IP addresses in the subnet's CIDR. Make sure that the credentials that you give the installer are authorized to do this! The installer will also apply a tag to the subnet. Lastly, the installer will not create a router, so make sure to do this if you need a router to attach floating IP addresses to the network you want to install the nodes onto. Note that for now, this feature does not fully support provider networks.
+In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published, and the Ingress and API ports are created. By default, the installer creates a network and subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the UUID of an existing OpenStack subnet. To use this feature there are a few requirements:
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:153]: Optional: 

```suggestion
In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published. The Ingress and API ports are created on this subnet, too. By default, the installer creates a network and subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the UUID of an existing OpenStack subnet. To use this feature, there are a few requirements:
```

Nitpicking:

At the end, "There are" is a little mushy to me. Something like "The target environment must meet the following requirements" seems closer.

File: docs/user/openstack/customization.md [hunk 19]
```diff
@@ -150,9 +150,18 @@ platform:
...
+* The subnet that is used by `machinesSubnet` has DHCP enabled.
+* The CIDR of `machinesSubnet` matches the CIDR of `networks.machineNetwork`.
+* The installer user must have permission to create ports on this network, including ports with fixed IP addresses.
+
+You should also be aware of the following limitations:
+
+* If you plan to install with floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`.
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:161]:
```suggestion
* If you plan to install a cluster that uses floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`.
```

File: docs/user/openstack/customization.md [hunk 20]
```diff
@@ -150,9 +150,18 @@ platform:
...
+* The installer will not create a private network or subnet for your OpenShift machines if the `machinesSubnet` is set in the `install-config.yaml`.
+* By default, the API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```
Comment (maxwelldb) [inline docs/user/openstack/customization.md:163]:
```suggestion
* By default, the API and Ingress VIPs use the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```

**Review by maxwelldb — APPROVED** [review]

File: docs/user/openstack/README.md (hunk 17 above)
Comment (iamemilio) [reply docs/user/openstack/README.md:508]: When defining a MachineSet with multiple networks, the `primarySubnet` value in the `providerSpec` must be set to the OpenStack subnet that you want the Kubernetes endpoints of nodes to be published on. For most use cases, this is the same subnet as the [machinesSubnet](./customization.md#cluster-scoped-properties) in the `install-config.yaml`.

How is this? I think it more accurately reflects the usage

Comment (maxwelldb) [reply docs/user/openstack/README.md:508]: Works for me. :+1:

Comment (iamemilio) [reply docs/user/openstack/README.md]: Take a look at what I pushed and see if it reads more clearly to you

File: docs/user/openstack/customization.md (hunk 18 above)
Comment (iamemilio) [reply docs/user/openstack/customization.md]: done

**Review by maxwelldb — APPROVED** [review]

**Comment by mandre** [comment]
/hold
Remove the `/hold` when we have an associated BZ.
